Sources/StreamChat/Workers/MessageUpdater.swift

```diff
@@ -66,14 +69,26 @@ class MessageUpdater: Worker {
                 return
             }
             
-            self.apiClient.request(endpoint: .deleteMessage(messageId: messageId)) { result in
+            self.apiClient.request(endpoint: .deleteMessage(messageId: messageId, hard: hard)) { result in
                 self.database.write({ session in
                     let messageDTO = session.message(id: messageId)
                     switch result {
-                    case .success:
-                        messageDTO?.localMessageState = nil
+                    case let .success(response):
+                        guard let cid = messageDTO?.channel?.cid else { return }
+
+                        let deletedMessage = try session.saveMessage(
+                            payload: response.message,
+                            for: ChannelId(cid: cid),
+                            syncOwnReactions: false
+                        )
+                        deletedMessage?.localMessageState = nil
+
+                        if hard, let message = deletedMessage {
+                            session.delete(message: message)
```

Comment:
what's the point of hard deleting from Core Data and marking messages as hard deleted?

Reply:
keep in mind that session.delete can cause crashes (we hold references to messages in a couple places)

Reply:
The reason the messages are also hard deleted from CoreData is that it is easier to manage. We also need the `isHardDelete` property to handle failed scenarios, to be able to retry the hard delete. We could just have `isHardDelete` and not actually delete it from Core Data but this would mean we needed to expose this property from the DTO to the model as well since we make the UI Logic as part of the Models. We have properties checking for `deletedAt != nil` and this will require changing this logic everywhere, adding unnecessary complexity. Deleting directly from CoreData makes it much easier to manage.